File: src/quixote_bot/handlers.py
import asyncio
import logging
import time

from aiogram import Dispatcher, F, Router
from aiogram.exceptions import TelegramRetryAfter
from aiogram.filters import Command, CommandObject, CommandStart
from aiogram.types import CallbackQuery, ErrorEvent, Message
from aiogram.utils.keyboard import InlineKeyboardBuilder

from .config import Settings
from .messages import (
    FAST_TRACK_WORDS,
    LANG_BUTTONS,
    LANG_OPTIONS,
    render_ai_unavailable,
    render_blacklist_list,
    render_blacklist_result,
    render_broadcast_confirm,
    render_categories,
    render_confirmation,
    render_error,
    render_fast_track_notification,
    render_fast_track_usage,
    render_max_questions,
    render_media_instruction,
    render_owner_notification,
    render_owner_only,
    render_portfolio,
    render_processing,
    render_rate_limit,
    render_unauthorized,
    render_unread_header,
    render_unread_item,
    render_welcome,
)
from .openrouter import OpenRouterClient, OpenRouterError
from .qualification import (
    CONVERSATION_TTL_SECONDS,
    SYSTEM_PROMPT,
    analyze_with_retry,
    build_user_context,
    fallback_analysis,
    get_next_question,
    parse_context,
    serialize_context,
    should_ask_question,
)
from .storage import Storage

logger = logging.getLogger(__name__)

# Telegram rejects messages over 4096 characters; keep a safety margin.
MAX_MESSAGE_LENGTH = 3900


def is_owner(user_id: int, settings: Settings) -> bool:
    return user_id == settings.owner_id


def setup_handlers(dp: Dispatcher, settings: Settings, storage: Storage, ai_client: OpenRouterClient) -> None:
    router = Router()

    @dp.errors()
    async def on_error(event: ErrorEvent) -> None:
        logger.exception("Unhandled exception while processing update: %s", event.exception)
        message = event.update.message
        if message is None or not hasattr(message, "answer"):
            return
        try:
            user = storage.get_user(message.from_user.id) if message.from_user else None
            lang = user.language if user else "ru"
            await message.answer(render_error(lang))
        except Exception:
            logger.warning("Failed to deliver error notice to the user")

    async def send_fast_track(message: Message, lang: str, body: str) -> None:
        uid = message.from_user.id
        user = storage.get_user(uid)
        uname = user.username if user else None
        storage.clear_conversation(uid)
        storage.save_message(uid, body, "готовое ТЗ")
        try:
            await message.bot.send_message(
                settings.owner_id, render_fast_track_notification(uid, uname, body)
            )
        except Exception:
            logger.exception("Failed to deliver fast-track application for user %s", uid)
        else:
            # Keep unread on delivery failure so /new can pick the lead up.
            storage.mark_read(uid)
        await message.answer(render_confirmation(lang))

    @router.message(CommandStart())
    async def cmd_start(message: Message) -> None:
        uid = message.from_user.id
        uname = message.from_user.username
        storage.ensure_user(uid, uname)
        user = storage.get_user(uid)
        lang = user.language if user else "ru"

        builder = InlineKeyboardBuilder()
        for code, label in LANG_BUTTONS.items():
            builder.button(text=label, callback_data=f"lang:{code}")

        await message.answer(LANG_OPTIONS.get(lang, LANG_OPTIONS["ru"]), reply_markup=builder.as_markup())

    @router.callback_query(F.data.startswith("lang:"))
    async def on_language_selected(callback: CallbackQuery) -> None:
        lang = callback.data.split(":")[1]
        if lang not in LANG_BUTTONS:
            await callback.answer()
            return
        uid = callback.from_user.id
        uname = callback.from_user.username
        storage.ensure_user(uid, uname)
        storage.set_language(uid, lang)

        await callback.message.edit_text(render_welcome(lang))
        await callback.answer()

    @router.message(Command("portfolio"))
    async def cmd_portfolio(message: Message) -> None:
        if message.from_user is None:
            return
        user = storage.get_user(message.from_user.id)
        lang = user.language if user else "ru"
        await message.answer(render_portfolio(lang))

    @router.message(Command("new"))
    async def cmd_new(message: Message) -> None:
        if not is_owner(message.from_user.id, settings):
            await message.answer(render_owner_only())
            return

        unread = storage.get_unread()
        if not unread:
            await message.answer("Нет непрочитанных сообщений.")
            return

        header = render_unread_header(len(unread))
        chunks: list[tuple[str, set[int]]] = []
        current_text = header
        current_users: set[int] = set()
        for msg in unread:
            user = storage.get_user(msg.user_id)
            uname = user.username if user else None
            item = render_unread_item(msg.id, msg.user_id, uname, msg.text, msg.category)
            if current_users and len(current_text) + len(item) + 2 > MAX_MESSAGE_LENGTH:
                chunks.append((current_text, current_users))
                current_text = ""
                current_users = set()
            current_text = f"{current_text}\n\n{item}" if current_text else item
            current_users.add(msg.user_id)
        if current_users:
            chunks.append((current_text, current_users))

        # Mark messages read only after their chunk was actually delivered,
        # so a failed send leaves them for the next /new call.
        for chunk_text, user_ids in chunks:
            await message.answer(chunk_text)
            for uid in user_ids:
                storage.mark_read(uid)

    @router.message(Command("categories"))
    async def cmd_categories(message: Message) -> None:
        if not is_owner(message.from_user.id, settings):
            await message.answer(render_owner_only())
            return

        stats = storage.get_category_counts()
        await message.answer(render_categories(stats))

    @router.message(Command("broadcast"))
    async def cmd_broadcast(message: Message, command: CommandObject) -> None:
        if not is_owner(message.from_user.id, settings):
            await message.answer(render_owner_only())
            return

        text = (command.args or "").strip()
        if not text:
            await message.answer("Использование: /broadcast <сообщение>")
            return

        count = 0
        failed = 0
        for uid in storage.get_broadcast_targets():
            try:
                await message.bot.send_message(uid, text)
                count += 1
            except TelegramRetryAfter as e:
                await asyncio.sleep(e.retry_after)
                try:
                    await message.bot.send_message(uid, text)
                    count += 1
                except Exception:
                    failed += 1
                    logger.warning("Broadcast to %s failed after retry", uid)
            except Exception:
                failed += 1
                logger.warning("Broadcast to %s failed", uid)
            await asyncio.sleep(0.05)

        if failed:
            logger.warning("Broadcast finished: %d delivered, %d failed", count, failed)
        await message.answer(render_broadcast_confirm(count))

    @router.message(Command("blacklist"))
    async def cmd_blacklist(message: Message) -> None:
        if not is_owner(message.from_user.id, settings):
            await message.answer(render_owner_only())
            return

        parts = message.text.split(maxsplit=3)
        if len(parts) < 2:
            await message.answer(
                "Использование:\n"
                "/blacklist add <user_id> [причина]\n"
                "/blacklist remove <user_id>\n"
                "/blacklist list"
            )
            return

        action = parts[1].lower()

        if action == "list":
            await message.answer(render_blacklist_list(storage.get_blacklist()))
            return

        if len(parts) < 3:
            await message.answer("Укажите user_id.")
            return

        try:
            target_id = int(parts[2])
        except ValueError:
            await message.answer("user_id должен быть числом.")
            return

        reason = parts[3] if len(parts) > 3 else ""

        if action == "add":
            success = storage.blacklist_add(target_id, reason)
            await message.answer(render_blacklist_result("add", str(target_id), success))
        elif action == "remove":
            success = storage.blacklist_remove(target_id)
            await message.answer(render_blacklist_result("remove", str(target_id), success))
        else:
            await message.answer("Неизвестная команда. Используйте add/remove/list.")

    @router.message(Command("apply"))
    async def cmd_apply(message: Message, command: CommandObject) -> None:
        if message.from_user is None:
            return
        uid = message.from_user.id
        text = (command.args or "").strip()

        storage.ensure_user(uid, message.from_user.username)
        user = storage.get_user(uid)
        lang = user.language if user else "ru"

        if is_owner(uid, settings):
            return

        if storage.is_blacklisted(uid):
            await message.answer(render_unauthorized())
            return

        if not text:
            await message.answer(render_fast_track_usage(lang))
            return

        if not storage.check_rate_limit(uid, settings.rate_limit, settings.rate_window):
            await message.answer(render_rate_limit(lang))
            return

        await send_fast_track(message, lang, text)

    @router.message(F.content_type.in_({"photo", "video", "document", "voice", "video_note", "sticker", "animation"}))
    async def on_media(message: Message) -> None:
        if message.from_user is None:
            return
        uid = message.from_user.id
        user = storage.get_user(uid)
        lang = user.language if user else "ru"

        if is_owner(uid, settings):
            return

        if storage.is_blacklisted(uid):
            return

        await message.answer(render_media_instruction(lang))

    @router.message(F.text)
    async def on_text(message: Message) -> None:
        if message.from_user is None:
            return
        uid = message.from_user.id
        text = message.text.strip()

        if not text:
            return

        storage.ensure_user(uid, message.from_user.username)
        user = storage.get_user(uid)
        lang = user.language if user else "ru"

        if is_owner(uid, settings):
            return

        if storage.is_blacklisted(uid):
            await message.answer(render_unauthorized())
            return

        if not storage.check_rate_limit(uid, settings.rate_limit, settings.rate_window):
            await message.answer(render_rate_limit(lang))
            return

        # Fast track: a message starting with the magic word goes straight to
        # the owner as a ready application, with no AI call and no questions.
        first_word, _, rest = text.partition(" ")
        if first_word.strip(".,:!?;\"'()").lower() in FAST_TRACK_WORDS:
            body = rest.strip()
            if not body:
                await message.answer(render_fast_track_usage(lang))
                return
            await send_fast_track(message, lang, body)
            return

        conv = storage.get_conversation(uid)
        if conv and time.time() - conv.started_at > CONVERSATION_TTL_SECONDS:
            storage.clear_conversation(uid)
            conv = None

        # Accumulate the brief: the original message plus answers to
        # clarifying questions are classified together.
        context_messages = parse_context(conv.context) if conv else []
        context_messages.append(text)

        await message.answer(render_processing(lang))

        # The OpenRouter client is blocking (up to 30 s per attempt), so run
        # classification in a worker thread to keep the event loop responsive.
        try:
            analysis = await asyncio.to_thread(
                analyze_with_retry,
                lambda: ai_client.chat(SYSTEM_PROMPT, build_user_context(context_messages)),
            )
        except OpenRouterError as e:
            logger.warning("AI analysis failed for user %s: %s", uid, e)
            analysis = fallback_analysis()
            await message.answer(render_ai_unavailable(lang))

        storage.save_message(uid, text, analysis.category)

        if analysis.category == "спам":
            logger.info("User %s auto-blacklisted as spam", uid)
            storage.blacklist_add(uid, "автоматический спам")
            return

        questions_asked = conv.questions_asked if conv else 0

        if should_ask_question(analysis, questions_asked):
            question = get_next_question(questions_asked, lang)
            if question:
                storage.upsert_conversation(
                    uid, "asking", questions_asked + 1, serialize_context(context_messages)
                )
                await message.answer(question)
                return

        # Completed — send to owner
        storage.clear_conversation(uid)

        uname = user.username if user else None
        brief = "\n\n".join(context_messages)
        notification = render_owner_notification(uid, uname, brief, analysis, questions_asked)

        try:
            await message.bot.send_message(settings.owner_id, notification)
        except Exception:
            logger.exception("Failed to deliver lead notification for user %s", uid)
        else:
            # Keep unread on delivery failure so /new can pick the lead up.
            storage.mark_read(uid)

        client_reply = analysis.client_reply
        if client_reply:
            await message.answer(client_reply)
        else:
            fallback = render_max_questions(lang) if questions_asked else render_confirmation(lang)
            await message.answer(fallback)

    dp.include_router(router)

File: cf-worker/src/handlers.ts
// Порт src/quixote_bot/handlers.py на grammY: вся логика бота.

import { Bot, Context, InlineKeyboard } from "grammy";

import * as db from "./db";
import {
    FAST_TRACK_WORDS,
    LANG_BUTTONS,
    LANG_OPTIONS,
    buildUnreadChunks,
    renderAiUnavailable,
    renderBlacklistList,
    renderBlacklistResult,
    renderBroadcastConfirm,
    renderCategories,
    renderConfirmation,
    renderError,
    renderFastTrackNotification,
    renderFastTrackUsage,
    renderMaxQuestions,
    renderMediaInstruction,
    renderOwnerNotification,
    renderOwnerOnly,
    renderPortfolio,
    renderProcessing,
    renderRateLimit,
    renderUnauthorized,
    renderUnreadHeader,
    renderUnreadItem,
    renderWelcome,
} from "./messages";
import { OpenRouterError, chat } from "./openrouter";
import {
    CONVERSATION_TTL_SECONDS,
    SYSTEM_PROMPT,
    analyzeWithRetry,
    buildUserContext,
    fallbackAnalysis,
    getNextQuestion,
    parseContext,
    serializeContext,
    shouldAskQuestion,
} from "./qualification";
import type { Env } from "./types";

const RATE_LIMIT = 10;
const RATE_WINDOW = 300;

// Порядок команд в grammY важен: команды зарегистрированы до message:text,
// как в aiogram-версии.

function isOwner(ctx: Context, ownerId: number): boolean {
    return ctx.from?.id === ownerId;
}

// Возвращает тело быстрой заявки, "" если текста нет, null если не fast-track.
export function matchFastTrack(text: string): string | null {
    const idx = text.indexOf(" ");
    const first = (idx === -1 ? text : text.slice(0, idx))
        .replace(/^[.,:!?;"'()]+|[.,:!?;"'()]+$/g, "")
        .toLowerCase();
    if (!FAST_TRACK_WORDS.has(first)) return null;
    return idx === -1 ? "" : text.slice(idx + 1).trim();
}

// Аргументы команды: всё после первого токена ("/apply текст" → "текст").
function commandArgs(ctx: Context): string {
    const raw = ctx.message?.text ?? "";
    const parts = raw.trim().split(/\s+/);
    return parts.slice(1).join(" ").trim();
}

export function createBot(env: Env): Bot {
    const bot = new Bot(env.BOT_TOKEN);
    const dbi = env.DB;
    const ownerId = Number(env.OWNER_ID);

    async function getLang(userId: number): Promise<string> {
        const user = await db.getUser(dbi, userId);
        return user?.language ?? "ru";
    }

    async function deliverFastTrack(ctx: Context, lang: string, body: string): Promise<void> {
        const uid = ctx.from!.id;
        const user = await db.getUser(dbi, uid);
        await db.clearConversation(dbi, uid);
        await db.saveMessage(dbi, uid, body, "готовое ТЗ");
        let delivered = false;
        try {
            await ctx.api.sendMessage(
                ownerId,
                renderFastTrackNotification(uid, user?.username ?? null, body),
            );
            delivered = true;
        } catch (e) {
            console.error(`Failed to deliver fast-track application for user ${uid}:`, e);
        }
        // При ошибке доставки заявка остаётся непрочитанной и подхватится /new.
        if (delivered) await db.markRead(dbi, uid);
        await ctx.reply(renderConfirmation(lang));
    }

    bot.catch((err) => {
        console.error("Unhandled error while processing update:", err.error);
        const ctx = err.ctx;
        if (ctx) {
            getLang(ctx.from?.id ?? 0)
                .then((lang) => ctx.reply(renderError(lang)))
                .catch(() => console.warn("Failed to deliver error notice to the user"));
        }
    });

    bot.command("start", async (ctx) => {
        const uid = ctx.from!.id;
        await db.ensureUser(dbi, uid, ctx.from!.username ?? null);
        const lang = await getLang(uid);

        const keyboard = new InlineKeyboard();
        for (const [code, label] of Object.entries(LANG_BUTTONS)) {
            keyboard.text(label, `lang:${code}`).row();
        }
        await ctx.reply(LANG_OPTIONS[lang] ?? LANG_OPTIONS["ru"]!, { reply_markup: keyboard });
    });

    bot.callbackQuery(/^lang:(ru|en)$/, async (ctx) => {
        const lang = ctx.callbackQuery.data!.split(":")[1]!;
        const uid = ctx.from.id;
        await db.ensureUser(dbi, uid, ctx.from.username ?? null);
        await db.setLanguage(dbi, uid, lang);
        await ctx.editMessageText(renderWelcome(lang));
        await ctx.answerCallbackQuery();
    });

    // Некорректный callback-данные молча игнорируем (валидация языка выше).
    bot.callbackQuery(/^lang:/, (ctx) => ctx.answerCallbackQuery());

    bot.command("portfolio", async (ctx) => {
        if (!ctx.from) return;
        const lang = await getLang(ctx.from.id);
        await ctx.reply(renderPortfolio(lang));
    });

    bot.command("apply", async (ctx) => {
        if (!ctx.from) return;
        const uid = ctx.from.id;
        const text = commandArgs(ctx);

        await db.ensureUser(dbi, uid, ctx.from.username ?? null);
        const lang = await getLang(uid);

        if (isOwner(ctx, ownerId)) return;
        if (await db.isBlacklisted(dbi, uid)) {
            await ctx.reply(renderUnauthorized());
            return;
        }
        if (!text) {
            await ctx.reply(renderFastTrackUsage(lang));
            return;
        }
        if (!(await db.checkRateLimit(dbi, uid, RATE_LIMIT, RATE_WINDOW))) {
            await ctx.reply(renderRateLimit(lang));
            return;
        }
        await deliverFastTrack(ctx, lang, text);
    });

    bot.command("new", async (ctx) => {
        if (!isOwner(ctx, ownerId)) {
            await ctx.reply(renderOwnerOnly());
            return;
        }
        const unread = await db.getUnread(dbi);
        if (unread.length === 0) {
            await ctx.reply("Нет непрочитанных сообщений.");
            return;
        }
        const items = [];
        for (const msg of unread) {
            const user = await db.getUser(dbi, msg.userId);
            items.push({
                userId: msg.userId,
                rendered: renderUnreadItem(msg.id, msg.userId, user?.username ?? null, msg.text, msg.category),
            });
        }
        const chunks = buildUnreadChunks(renderUnreadHeader(unread.length), items);
        // Помечаем прочитанным только после фактической отправки чанка:
        // неудачная отправка оставляет сообщения на следующий /new.
        for (const chunk of chunks) {
            await ctx.reply(chunk.text);
            for (const uid of [...new Set(chunk.userIds)]) {
                await db.markRead(dbi, uid);
            }
        }
    });

    bot.command("categories", async (ctx) => {
        if (!isOwner(ctx, ownerId)) {
            await ctx.reply(renderOwnerOnly());
            return;
        }
        await ctx.reply(renderCategories(await db.getCategoryCounts(dbi)));
    });

    bot.command("broadcast", async (ctx) => {
        if (!isOwner(ctx, ownerId)) {
            await ctx.reply(renderOwnerOnly());
            return;
        }
        const text = commandArgs(ctx);
        if (!text) {
            await ctx.reply("Использование: /broadcast <сообщение>");
            return;
        }
        let count = 0;
        let failed = 0;
        for (const uid of await db.getBroadcastTargets(dbi)) {
            try {
                await ctx.api.sendMessage(uid, text);
                count++;
            } catch (e) {
                failed++;
                console.warn(`Broadcast to ${uid} failed:`, e);
            }
        }
        if (failed > 0) console.warn(`Broadcast finished: ${count} delivered, ${failed} failed`);
        await ctx.reply(renderBroadcastConfirm(count));
    });

    bot.command("blacklist", async (ctx) => {
        if (!isOwner(ctx, ownerId)) {
            await ctx.reply(renderOwnerOnly());
            return;
        }
        const parts = (ctx.message?.text ?? "").trim().split(/\s+/);
        if (parts.length < 2) {
            await ctx.reply(
                "Использование:\n/blacklist add <user_id> [причина]\n/blacklist remove <user_id>\n/blacklist list",
            );
            return;
        }
        const action = parts[1]!.toLowerCase();

        if (action === "list") {
            await ctx.reply(renderBlacklistList(await db.getBlacklist(dbi)));
            return;
        }
        if (parts.length < 3) {
            await ctx.reply("Укажите user_id.");
            return;
        }
        const targetId = Number(parts[2]);
        if (!Number.isInteger(targetId)) {
            await ctx.reply("user_id должен быть числом.");
            return;
        }
        const reason = parts.slice(3).join(" ");

        if (action === "add") {
            const success = await db.blacklistAdd(dbi, targetId, reason);
            await ctx.reply(renderBlacklistResult("add", String(targetId), success));
        } else if (action === "remove") {
            const success = await db.blacklistRemove(dbi, targetId);
            await ctx.reply(renderBlacklistResult("remove", String(targetId), success));
        } else {
            await ctx.reply("Неизвестная команда. Используйте add/remove/list.");
        }
    });

    bot.on(
        ["message:photo", "message:video", "message:document", "message:voice", "message:video_note", "message:sticker", "message:animation"],
        async (ctx) => {
            if (!ctx.from) return;
            const lang = await getLang(ctx.from.id);
            if (isOwner(ctx, ownerId)) return;
            if (await db.isBlacklisted(dbi, ctx.from.id)) return;
            await ctx.reply(renderMediaInstruction(lang));
        },
    );

    bot.on("message:text", async (ctx) => {
        if (!ctx.from) return;
        const uid = ctx.from.id;
        const text = ctx.message.text.trim();
        if (!text) return;

        await db.ensureUser(dbi, uid, ctx.from.username ?? null);
        const lang = await getLang(uid);

        if (isOwner(ctx, ownerId)) return;
        if (await db.isBlacklisted(dbi, uid)) {
            await ctx.reply(renderUnauthorized());
            return;
        }
        if (!(await db.checkRateLimit(dbi, uid, RATE_LIMIT, RATE_WINDOW))) {
            await ctx.reply(renderRateLimit(lang));
            return;
        }

        // Быстрая заявка: слово ЗАКАЗ/ORDER в начале — напрямую владельцу,
        // без ИИ и без вопросов.
        const fastBody = matchFastTrack(text);
        if (fastBody !== null) {
            if (!fastBody) {
                await ctx.reply(renderFastTrackUsage(lang));
                return;
            }
            await deliverFastTrack(ctx, lang, fastBody);
            return;
        }

        // Накапливаем бриф: ответы на уточняющие вопросы классифицируются
        // вместе с исходным сообщением.
        let conv = await db.getConversation(dbi, uid);
        if (conv && Date.now() / 1000 - conv.startedAt > CONVERSATION_TTL_SECONDS) {
            await db.clearConversation(dbi, uid);
            conv = null;
        }
        const contextMessages = parseContext(conv?.context ?? null);
        contextMessages.push(text);

        await ctx.reply(renderProcessing(lang));

        let analysis;
        try {
            analysis = await analyzeWithRetry(() =>
                chat(env, SYSTEM_PROMPT, buildUserContext(contextMessages)),
            );
        } catch (e) {
            if (!(e instanceof OpenRouterError)) throw e;
            console.warn(`AI analysis failed for user ${uid}: ${e.message}`);
            analysis = fallbackAnalysis();
            await ctx.reply(renderAiUnavailable(lang));
        }

        await db.saveMessage(dbi, uid, text, analysis.category);

        if (analysis.category === "спам") {
            console.info(`User ${uid} auto-blacklisted as spam`);
            await db.blacklistAdd(dbi, uid, "автоматический спам");
            return;
        }

        const questionsAsked = conv?.questionsAsked ?? 0;

        if (shouldAskQuestion(analysis, questionsAsked)) {
            const question = getNextQuestion(questionsAsked, lang);
            if (question) {
                await db.upsertConversation(
                    dbi,
                    uid,
                    "asking",
                    questionsAsked + 1,
                    serializeContext(contextMessages),
                );
                await ctx.reply(question);
                return;
            }
        }

        // Заявка готова — отправляем владельцу.
        await db.clearConversation(dbi, uid);
        const user = await db.getUser(dbi, uid);
        const brief = contextMessages.join("\n\n");
        const notification = renderOwnerNotification(uid, user?.username ?? null, brief, analysis, questionsAsked);

        let delivered = false;
        try {
            await ctx.api.sendMessage(ownerId, notification);
            delivered = true;
        } catch (e) {
            console.error(`Failed to deliver lead notification for user ${uid}:`, e);
        }
        // При ошибке доставки остаётся непрочитанным — подхватит /new.
        if (delivered) await db.markRead(dbi, uid);

        if (analysis.clientReply) {
            await ctx.reply(analysis.clientReply);
        } else {
            const fallback = questionsAsked ? renderMaxQuestions(lang) : renderConfirmation(lang);
            await ctx.reply(fallback);
        }
    });

    return bot;
}
